### PR TITLE
Store invalid name in the name error

### PR DESCRIPTION
**Pull request by bjfish**

### CODE REVIEW

File: src/main/java/org/truffleruby/core/exception/CoreExceptions.java [hunk 1]
```diff
@@ -573,6 +578,11 @@ public DynamicObject nameErrorUnknownIdentifier(TruffleObject receiver, Object n
 
     @TruffleBoundary
     public DynamicObject nameError(String message, Object receiver, String name, Node currentNode) {
+        return nameError(message, receiver, context.getSymbolTable().getSymbol(name), currentNode);
+    }
+
+    @TruffleBoundary
+    public DynamicObject nameError(String message, Object receiver, Object name, Node currentNode) {
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/exception/CoreExceptions.java]: Could you be type `name` as DynamicObject and assert it's a Symbol?

File: src/main/java/org/truffleruby/cext/CExtNodes.java [hunk 2]
```diff
@@ -1074,9 +1074,9 @@ public Object debug(Object... objects) {
     public abstract static class HiddenVariableGetNode extends CoreMethodArrayArgumentsNode {
 
         @Specialization(guards = "isRubySymbol(name)")
-        public Object hiddenVariableGet(DynamicObject object, DynamicObject name,
+        public Object hiddenVariableGet(VirtualFrame frame, DynamicObject object, DynamicObject name,
                 @Cached("createObjectIVarGetNode()") ObjectIVarGetNode iVarGetNode) {
-            return iVarGetNode.executeIVarGet(object, name);
+            return iVarGetNode.executeIVarGet(frame, object, name);
```
Comment (eregon) [inline src/main/java/org/truffleruby/cext/CExtNodes.java]: Could you rebase on top of #812 once it's merged?
That should make the diff smaller.

File: src/main/java/org/truffleruby/language/objects/CheckInstanceVariableNameNode.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public abstract String executeCheck(VirtualFrame frame, DynamicObject object, Object name);
+
+    @Specialization
+    public String check(VirtualFrame frame, DynamicObject object, Object name,
+                        @Cached("create()") NameToJavaStringNode nameToJavaStringNode) {
+        String nameAsString = nameToJavaStringNode.executeToJavaString(frame, name);
+        if (!Identifiers.isValidInstanceVariableName(nameAsString)) {
+            throw new RaiseException(getContext().getCoreExceptions().nameErrorInstanceNameNotAllowable(
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/language/objects/CheckInstanceVariableNameNode.java]: This should have a branch profile.

File: src/main/java/org/truffleruby/core/exception/CoreExceptions.java (hunk 1 above)
Comment (bjfish) [reply src/main/java/org/truffleruby/core/exception/CoreExceptions.java]: In this spec, the name is a ruby String:
https://github.com/ruby/ruby/blob/5a8a2706d33849cf42cd56714f519108a4ededfc/spec/ruby/core/exception/name_spec.rb#L46

**Comment by bjfish** [comment]
I think I'll close this for now until I make progress on handling more scenarios. I'm finding more scenarios like:
```
      invalid_ivar_obj = Object.new
      def invalid_ivar_obj.to_str
        "to_str_var"
      end
      -> {
        Object.new.instance_variable_get(invalid_ivar_obj)
      }.should raise_error(NameError) {|e| e.name.should == "to_str_var" }

```

**Pull request closed without merging**